spec/abstract/SchedContext_A.thy

```diff
@@ -195,72 +190,76 @@ where
                         \<lparr>r_time = r_time (hd (tl refills)), r_amount = r_amount (hd (tl refills)) + usage\<rparr>]
    od"
 
+function
+  MIN_BUDGET_merge :: "refill list \<Rightarrow> refill list"
+where
+  "MIN_BUDGET_merge refills
+    = (if r_amount (hd refills) < MIN_BUDGET
+       then let new_hd = \<lparr>r_time = r_time (hd (tl refills)) - r_amount (hd refills),
+                          r_amount = r_amount (hd refills) + r_amount (hd (tl refills))\<rparr>
+            in MIN_BUDGET_merge (new_hd # (tl (tl refills)))
+       else refills)"
+by pat_completeness auto
+
```

Comment:
This function is strictly partial, but every refill list (ignoring word addition overflow issues) which has nonzero sum is an element of the domain. Since `MIN_BUDGET` is nonzero, and we are interested only in refill lists where the (`unat`) sum of the refill amounts is a least `MIN_BUDGET`, this is fine. My current thought is to prove a lemma saying that all refill lists where the `unat` sum of the r_amounts is at least `MIN_BUDGET` are in the domain. Then proving lemmas about this function (for instance, that this function will actually return a list with at least `MIN_BUDGET` in the head) is "straightforward"